Express integration tests with different error handling

esclapes opened the pull request:
Hey @hharnisc, 

This PR includes some tests to showcase an issue that (I think) I found when reviewing how `error.handled` property is set on the current `RPCClient` after your comment [here](https://github.com/bufferapp/micro-rpc-client/pull/1). No need to merge it.

I found that on an server unhandled exception, the client will behave inconsistently depending on the error handling logic in the server. The `test-apps.js` file shows 3 error handling alternatives:

1. **No custom error handler**: express default error handling kicks in, which means the response is _not_ JSON. That means the client will indeed throw an _unhandled error_. However it will be a _FetchError_ instead of the original server error which hides the underlying issue.
2. **A custom error handler that returns a `{error: err.message}`**: with this one the client will indeed throw the error with the right message, but it will consider it a _handled error_
3. **A custom error handler that returns a `{err: err.message}`**: this one will trick the client into thinking it got a valid result and resolve to `undefined`

If this is indeed an issue, I would suggest a two part solution:

* Provide an error handling middleware in this new `buffer-rpc` package, and document how to use it in express.
* Update the client code to handle the server error handling approach.

Looking forward to hearing your take on this, as I might be missing some context here too :smile:

On a side note. How do you feel about bundling the new `Client` implementation in this repository? I find it would make it easier to do integration testing and it would make it easier to keep server and client in sync as they evolve.

esclapes commented:
cc @philippemiguet

hharnisc commented on test/integration/express.test.js:
```diff
@@ -0,0 +1,58 @@
+const Client = require('micro-rpc-client')
+const getPort = require('get-port')
```
Great idea to integration tests @esclapes :raised_hands: 

It looks like creating a test service to run RPC requests against is spun up a different way than the rest of the tests.

There's a few examples using the utils with the `test-listen` library: https://github.com/bufferapp/buffer-rpc/blob/master/test/rpc.test.js#L228

It accomplishes the same thing is this code :smile: 

So in effect all of the tests are integration tests -- we're just missing an integration test with the RPC library :wink:

esclapes commented:
Hey @hharnisc, when I read now the original issue body I am thinking I missed the opportunity to focus the discussion. If I understand well the `Error.handled` property:

* it should be false when the error in the server was _unexpected_.
* the error message itself should match the unexpected error in the server.

With the current client and with three different error handler implementations in the server, I could not get these two constrains to be satisfied:

This is the outcome I got after running `npm run test client` with this branch.

![image](https://user-images.githubusercontent.com/829971/45939527-8975bf00-bfd3-11e8-8371-e980b1ec49ea.png)

I may be missing something here, so I would love to take your take on this.

esclapes commented:
Also, please let me know if you feel I should update the issue body to make it more clear.

esclapes commented:
Hey @philippemiguet the latest changes Harrison added to both RPC client and server, fixed this issue when using the available `errorMiddleware` in `buffer-rpc`

I have updated the tests in this PR (force pushed it :sweat_smile:) to remove the other cases and check for some other integration potins with the client. How do you feel about having this integration tests merged into the pacakge?

![image](https://user-images.githubusercontent.com/829971/49498975-0f807280-f86d-11e8-9e14-5e4cbfbb9874.png).

philippemiguet commented:
@esclapes the more tests the merrier 😂 😄 👍

esclapes commented:
>  the more tests the merrier joy smile +1

Your wish is my command :grin: 

![image](https://user-images.githubusercontent.com/829971/49499354-1065d400-f86e-11e8-96f3-cd32b44893eb.png)
